### PR TITLE
Add linked account creation from the Discord bot

### PR DESCRIPTION
**Pull request by daanbreur**
Implements #300

### CODE REVIEW

**Comment by daanbreur** [comment]
Discovered a little problem. Now you could make multiple invite codes that point to the same user and then use them seperately and have the database error out. Possible to return the old code if this one already exists.

**Comment by daanbreur** [comment]
Another problem, with the convential link command the command will automatically sync permissions for all ctfs after linking. Currently I have no idea how this can be implemented.

**Comment by daanbreur** [comment]
> Discovered a little problem. Now you could make multiple invite codes that point to the same user and then use them seperately and have the database error out. Possible to return the old code if this one already exists.

Discord bot now checks if there is already a invitation token for the Discord id. If there is, it returns this token instead of making a new one. The database can now be put back to unique and this eliminates the issue of invitation links for the same account causing errors and issues.

image

**Comment by daanbreur** [comment]
The message when the functionality has not been enabled.
image

---
Also added the check if you have the required role to create an account.

Message when you DO NOT have the role.
image

Message when you DO have the role
image

**Comment by daanbreur** [comment]
Currently stuck at the "account syncing" part. Using the already existing `/link` your account would, after linking, automatically get all roles and access to ctfs. However with the stack we created we are kinda driven into a corner. 

Within the discord-hooks we can hook into the `RegisterWithToken` graphql query that is being used. There is a problem that arose tho, within this context we are the user `user_postgraphile`, except this isn't entirely true. The context we are in uses the user we used to call the GraphQL query, since we are on the login we aren't any user and we are `user_anonymous`. This complicates our situation a bit.

For the discord role sync we need a few things, the ctfnote user-id of the just created user and that this user exists in the database.

File: api/src/discord/commands/register.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+async function getProfileUrl() {
+  if (config.pad.domain == "") return null;
+
+  const ssl = config.pad.useSSL == "false" ? "" : "s";
+
+  return `http${ssl}://${config.pad.domain}/#/user/settings`;
+}
+
+async function createAccountLogic(
```
Comment (daanbreur) [inline api/src/discord/commands/register.ts:33]: function name should be `registerLogic` to match with the command name.

File: api/src/plugins/discordHooks.ts [hunk 2]
```diff
@@ -279,6 +283,30 @@ const discordMutationHook = (_build: Build) => (fieldContext: Context<any>) => {
...
+        for (let i = 0; i < ctfs.length; i++) {
+          await changeDiscordUserRoleForCTF(userId, ctfs[i], "add").catch(
+            (err) => {
+              console.error("Error while adding role to user: ", err);
+            }
+          );
+        }
+      }, 2000);
+    }
+
```
Comment (daanbreur) [inline api/src/plugins/discordHooks.ts:309]: approved, should be made better when the Discord part is rewritten if possible.

**Review by JJ-8 — APPROVED** [review]
I tried to test everything and it works as expected. I made some changes to the code and it looks all good now. I prefer to delay the merging to next Saturday just in case due to time constrains on my side.

**Comment by daanbreur** [comment]
> I tried to test everything and it works as expected. I made some changes to the code and it looks all good now. I prefer to delay the merging to next Saturday just in case due to time constrains on my side.

All looks good, should reimplement this some day when we change the backend to something that allows for easier hooking.

LGTM (on Saturday)

**Pull request merged**